proposals/accepted/000-ast-parser-libs.rst

```diff
@@ -0,0 +1,344 @@
+
+It is the opinion of the authors of this proposal that should an independent AST parser libraries be maintained upstream with GHC, the costs saved for downstream developers should *greatly* exceed any costs incurred by GHC developers.
+The goal is thus *not* to simply shift a burden from one group of community members to another, but create a positive-sum outcome where there is far less busywork and more flourishing tooling than before.
+
+.. [#example-ghc-lib-parser-users]
+  Today for example, notable users include
+  HLint_,
+   `ormolu <https://hackage.haskell.org/package/ormolu>`_,
+   `ghcide <https://hackage.haskell.org/package/ghcide>`_,
+   `hls-hlint-plugin <https://hackage.haskell.org/package/hls-hlint-plugin>`_,
+   `hindent <https://hackage.haskell.org/package/hindent>`_,
+   and
+   `stylish-haskell <https://hackage.haskell.org/package/stylish-haskell>`_.
+
+Trees that grow
+---------------
+
+As we can see, each of these prior two attempts did one of the two things right, and correspondingly met one of our two criteria.
+There is, however, a third project, that over the years has aimed to allow us to finally hit both criteria: "Trees that grow".
+The name comes from `this paper <https://www.microsoft.com/en-us/research/uploads/prod/2016/11/trees-that-grow.pdf>`_.
+There are also
+`some GHC Wiki pages <https://gitlab.haskell.org/ghc/ghc/-/wikis/implementing-trees-that-grow>`_,
+and a `GHC Issue Label <https://gitlab.haskell.org/ghc/ghc/-/issues/?label_name%5B%5D=TTG>`_ for it.
+
+The goal of the Trees that Grow paper was to allow creating variants of Haskell AST to more faithfully capture the input and output of each compilation pass, and also the ``template-haskell`` library. [#intra]_
+It presents these data types:
+
+.. code-block:: haskell
+
+  data Component = Compiler Pass | TemplateHaskell
+
+  data Pass = Parser | Renamer | TypeChecker
+
+The idea that they are "promoted" via ``DataKinds``, and then type families used in the AST will have instances for these promoted values.
+This allows those consumers to "adjust" the AST for their purpose.
+
+The Trees That Grow project is now 6 years old, and has met great success in avoiding partiality in the compiler, "making illegal states unrepresentable" as many Haskellers would put it.
+But progress on `reducing AST & parser dependencies <https://gitlab.haskell.org/ghc/ghc/-/issues/19932>`_ has been less easily forthcoming.
+We have separated out the modules defining the AST under ``Language.Haskell.Syntax.*`` we wish to split out, and we have tests to track progress reducing their deps, and the parser's deps.
+But progress is unsteady and unpredictable.
+
+The basic problem is that the benefits don't actually kick in until the deps are *all* gone, and the code is actually separated out.
+Partial progress isn't really directly useful to anyone, and these counters just scoreboard by which we hope to get closer to the end goal.
```

Comment:
What does "these counters just scoreboard by which" mean?

Reply:
in this paragraph I am trying to contrast the smooth function of module count hopefully going down with the step function of "benefits only kick in when count hits 0". It's good to track getting closer to 0, but important to remember partial progress != partial benefit.

Perhaps this whole paragraph is confusing and should be reworded?

Reply:
I think what threw me was the lack of the words "are" and "a" surrounding "just" :-)

```suggestion
Partial progress isn't really directly useful to anyone, and these counters are just a scoreboard by which we hope to get closer to the end goal.
```